### DISCUSSION TITLE
Does AdaptivePoolingAllocator use more memory than PooledByteBufAllocator?

### DISCUSSION THREAD

**Discussion by He-Pin** [netty/netty#14135] [General]
Env:
1. Java 21 the latest version with jemalloc instead of glibc.
2. Netty 4.1.110.final 
3. With 16 cores 32GB box, 22GB max heap size + 5GB max direct memory size  
4. A Netty proxy application, event loop been shared.
5. `Allocater.buffer` been called from 2 thread pools:  `a: Netty's EventLoopGroup with size 32 max, b: a OrderredTaskRunner with size 128 max, been called from Platform Thread only, no virtual thread been used during the stress testing.
7. A stress testing HTTP Request/Response at 1000qps with payload 300KB, the application is a proxy application, where it handle chunks , parsering to json, change the content , and send to clients.

it was : 
- before:  `public static final ByteBufAllocator BYTE_BUF_ALLOCATOR = PooledByteBufAllocator.DEFAULT;` 
- after   :  `public static final ByteBufAllocator BYTE_BUF_ALLOCATOR = new AdaptiveByteBufAllocator(true);` 

and the `BYTE_BUF_ALLOCATOR` is been shared across the application.

The stress testing is from 1AM to 11 AM, but the memory of the hosts continues grow.

![image](https://github.com/netty/netty/assets/501740/67018826-bc7e-4ad8-9793-76221bf04267)

The host memory usage is above:

java opts:
```
admin      2646  538 52.6 28951692 17673456 ?   Sl   19:18 345:09 /opt/taobao/java/bin/java -Dspring.profiles.active=staging -server -Xms22937m -Xmx22937m -XX:+UseG1GC -XX:InitiatingHeapOccupancyPercent=40 -XX:-G1UseAdaptiveIHOP -XX:G1HeapRegionSize=32m --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens java.base/jdk.internal.misc=ALL-UNNAMED --add-exports java.base/sun.net.www.protocol.https=ALL-UNNAMED --add-exports java.base/sun.util.calendar=ALL-UNNAMED --add-exports java.base/java.lang.reflect=ALL-UNNAMED --add-opens java.base/sun.net.util=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.math=ALL-UNNAMED --add-opens java.base/sun.security.action=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens java.sql/java.sql=ALL-UNNAMED --add-opens java.base/java.util.regex=ALL-UNNAMED -XX:MetaspaceSize=512m -XX:MaxMetaspaceSize=512m -XX:MaxDirectMemorySize=6g -XX:SurvivorRatio=10 -XX:+ExplicitGCInvokesConcurrent -Dsun.rmi.dgc.server.gcInterval=2592000000 -Dsun.rmi.dgc.client.gcInterval=2592000000 -XX:ParallelGCThreads=16 -Xlog:gc*:/home/admin/logs/gc.log:time -XX:+EnableDynamicAgentLoading -XX:+TieredCompilation -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/home/admin/logs/java.hprof -Djava.awt.headless=true -Dsun.net.client.defaultConnectTimeout=10000 -Dsun.net.client.defaultReadTimeout=30000 -DJM.LOG.PATH=/home/admin/logs -DJM.SNAPSHOT.PATH=/home/admin/snapshots -Dfile.encoding=UTF-8 -Dhsf.publish.delayed=true -Dhsf.cs.delay.time=0 -Dbp.enable=true -Dproject.name=ultra-mtop -Dpandora.boot.wait=true -Dlog4j.defaultInitOverride=true -Dserver.port=7001 -Dmanagement.port=7002 -Dmanagement.server.port=7002 -Djdk.tracePinnedThreads=full -Djdk.virtualThreadScheduler.parallelism=128 -Djdk.virtualThreadScheduler.maxPoolSize=2048 -Djdk.virtualThreadScheduler.minRunnable=32 -Ddisable.pandora.mtop-uncenter=true -Dmtop.plugin=DISABLE -agentlib:jdwp=transport=dt_socket,address=*:8000,server=y,suspend=n -Dpandora.location=/home/admin/ultra-mtop/target/taobao-hsf.sar -classpath /home/admin/ultra-mtop/target/ultra-mtop -Dapp.location=/home/admin/ultra-mtop/target/ultra-mtop -Djava.endorsed.dirs= -Djava.io.tmpdir=/home/admin/ultra-mtop/.default/temp com.taobao.pandora.boot.loader.SarLauncher
```

Background:
We are trying to make use of the `AdaptiveByteBufAllocator` for 1. better performance 2. better virtual thread integration. 

Sorry for not having a heap dump to share.

**Comment by chrisvest** [comment]
Each of the 32 event loops will have their own magazine (these magazines have `shareable = false`).
The OrderedTaskRunner sounds like it'll go to the magazine array - it'd be interesting to know the size of this array.
Once we know how many magazines there are, it'd be interesting to know what their `localPrefChunkSize` is.

Another thing worth investigating are the `Chunk` objects, and their `refCntUp-refCntDown` deltas.
If there are lots of them with small deltas, like 1 or 2, then maybe the anti-generational hypothesis isn't holding so well for your app.
If some of them have a delta of zero, but the delegate `ByteBuf` hasn't been deallocated, then that could indicate a bug that produces leaks.

**Reply by franz1981** [reply]
We can have different options here: if they are using the common FJ pool by default it can use the same amount of threads than cores.
Which means that the Virtual Thread pool compete with Netty event loops for the resources, which is a suboptimal configuration which requires to:
1. Size them differently to not compete on cores or
2. Allow them to share the same magazines of the  event loop by disabling the event loops ones

**Reply by franz1981** [reply]
Regardless @chrisvest some time ago I have proposed a change for the original Netty pooled allocator which prove to be very complex to be implemented i.e. soft limits.
This could be revived for this new pool i.e. a timer thread which evaluate for how long a shared magazine isn't used and reap them (starting from the shared queues first) till specified soft limits (which could be 0 too!). This won't harm event loop threads but can reduce dramatically the memory footprint, if combined with JDK malloc trimming (@nitsanw love it), wdyt?

That would make this new pool the preferred choice for containers given that on scale to 1 can just bring down RSS to the bare minimum.

**Reply by He-Pin** [reply]
I have not looked into the implementation in detail，but when the stress is gone, qps ↓， the allocator should clean/free up some memory, like the `CompositeByteBuf#discardReadComponents`.

**Reply by franz1981** [reply]
>  the allocator should clean/free up some memory

what we are observing here? the RSS? the counters from the pool itself?  heap usage?

**Reply by chrisvest** [reply]
I think we should figure out why there is increased memory usage for this application, before thinking up solutions. I wrote a list of things to investigate, in https://github.com/netty/netty/discussions/14135#discussioncomment-9832966.

**Comment by franz1981** [comment]
any news @He-Pin ? or any new info about this?

re 

> what we are observing here? the RSS? the counters from the pool itself? heap usage?

**Reply by He-Pin** [reply]
Sorry for not updating this, the RSS memory is  https://en.wikipedia.org/wiki/Resident_set_size

**Reply by franz1981** [reply]
I know what RSS mean, I wanted to understand if it was the metric reported/under scrutiny

**Reply by He-Pin** [reply]
Yes, The RSS memory is the JVM processor's, this API Gateway is where we translate our Taobao APP to the Taobao English version, you can try it too, download the Taobao app in the app store and change location to US/SEG, etc.

I captured that metric picture before we were stressing before we went into production, sorry for not having time to stress it again with the new AdaptiveByteBufAllocator :( But because it's an API Gateway, the byteBufs' size is changeful.

**Reply by He-Pin** [reply]
As we are using the `jemalloc` with the AJDK, so I think the AdaptiveByteBufAllocator may have some allocation issues, eg, not fully utilizing small chunks, maybe needing something like defragmentation, I have not looked into much of the detail, sorry for that.
